File: C++/validate_binary_search_tree.cpp
// no. 98
#include "tree_node.h"
#include <string>
using namespace std;

class Solution {
public:
  bool ans = true;
  int get_extreme(TreeNode *root, string mode) {
    int left = root->val;
    int right = root->val;
    if (root->left) {
      left = get_extreme(root->left, mode);
      if ((mode == "max") & (left >= root->val)) {
        ans = false;
      }
    }
    if (root->right) {
      right = get_extreme(root->right, mode);
      if ((mode == "min") & (right <= root->val)) {
        ans = false;
      }
    }
    if (mode == "max") {
      return max(left, right);
    } else {
      return min(left, right);
    }
  }
  bool isValidBST(TreeNode *root) {
    get_extreme(root, "max");
    if (ans) {
      get_extreme(root, "min");
    }
    return ans;
  }
};

File: python/validate_binary_search_tree.py
# no. 98
from utils import TreeNode
from typing import Optional

class Solution:
    def __init__(self):
        self.ans = True
    def get_extreme(self, root, mode):
        left = root.val
        right = root.val
        if root.left:
            left = self.get_extreme(root.left, mode)
            if mode == 'max' and left >= root.val:
                self.ans = False
        if root.right:
            right = self.get_extreme(root.right, mode)
            if mode == 'min' and right <= root.val:
                self.ans = False
        if mode == 'max':
            return max(left, right)
        elif mode == 'min':
            return min(left, right)
            
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.get_extreme(root, 'max')
        if self.ans:
            self.get_extreme(root, 'min')
        return self.ans

